lib/src/parameters.dart

```diff
@@ -0,0 +1,64 @@
+import 'dart:convert';
+import 'package:collection/collection.dart';
+import 'package:http_parser/http_parser.dart';
+
+typedef Map<String, dynamic> GetParameters(String contentType, String body, Uri tokenEndpoint);
+
+void validate(bool condition, String message) {
+  if (condition) return;
+  throw new FormatException(message);
+}
+
+/// Parses parameters from a response with either a JSON or URL-encoded body.
+Map<String, dynamic> parseJsonOrUrlEncodedParameters(
+    String contentTypeString, String body, Uri tokenEndpoint) {
+
+  var contentType =
+      contentTypeString == null ? null : new MediaType.parse(contentTypeString);
+
+  try {
+    // The spec requires a content-type of application/json, but some endpoints
+    // (e.g. Dropbox) serve it as text/javascript instead.
+    validate(
+        contentType != null &&
+            (contentType.mimeType == "application/json" ||
+                contentType.mimeType == "text/javascript" ||
+                contentType.mimeType == "application/x-www-form-urlencoded"),
+        'content-type was "$contentType", expected "application/json" or "application/x-www-form-urlencoded"');
+  } on FormatException catch(e) {
+    throw new FormatException('Invalid OAuth response for "$tokenEndpoint": '
+        '${e.message}.\n\n$body');
+  }
```

Comment:
The default `getParameters()` shouldn't do this wrapping. It should just throw a `FormatException` with the message; for example:

```dart
throw new FormatException('content-type was "$contentType", expected '
    '"application/json" or "application/x-www-form-urlencoded"');
```

and then let the caller wrap that in an exception that explains what the response was for and what the body was.

Reply:
👍